### ISSUE TITLE
Specify a custom DerivedData folder

### ISSUE THREAD

**Issue by bvirlet** [tuist/tuist#2712]
I haven't found anywhere a way to specify a custom DerivedData folder for tuist. Is there a way to do this? 

Thanks!

**Comment by luispadron** [comment]
Hi there, looks like we can provide a custom path, take a [look here](https://github.com/tuist/tuist/blob/149a442dbfa58d6ad1218a18b817b59be61cae8a/Sources/TuistSupport/Utils/DeveloperEnvironment.swift#L30)

```swift
if let customLocation = try? System.shared.capture("/usr/bin/defaults", "read", "com.apple.dt.Xcode IDECustomDerivedDataLocation") {
    location = AbsolutePath(customLocation.chomp())
} else {
    // Default location
    location = fileHandler.homeDirectory.appending(RelativePath("Library/Developer/Xcode/DerivedData/"))
}
```

**Comment by bvirlet** [comment]
Thanks, @luispadron. What is the recommended way to set this value before running `tuist generate`? Is there a way to pass it to Tuist without setting global defaults (

**Comment by fortmarek** [comment]
Hi @bvirlet - afaict you can change the derived data location globally in Xcode, it's not something saved in `.pbxproj`. If you change your path globally, the snippet @luispadron has sent will find the correct location.

Could you maybe expand a little bit about your use-case and what's not working exactly?

**Comment by bvirlet** [comment]
Sorry for not replying earlier.

I'm having problems with Swift PM binary artifacts not resolving properly and often need to clean the derived data folder. To do that in a deterministic way, I'd need to specify a custom derived data path for my workspace. I can't use a global setting because I'm using a Macstadium machine which can potentially run multiple builds in parallel and I don't want to affect other builds' derived data path. Instead, I'd like the derived data directory to be in the current build directory. 

Without touching the project file, I'd looking for something such as `DERIVED_DATA=somepath tuist generate`.

**Comment by fortmarek** [comment]
You can also change the custom derived data path globally to a relative path as per the image: 
![image](https://user-images.githubusercontent.com/9371695/120384801-ec908580-c326-11eb-82cb-5098152fe907.png). Does that fix your problem?

**Comment by bvirlet** [comment]
I don't think it will help. I will try to describe my scenario better :

On my CI machine (I don't have access to Xcode UI), I run the following:

```
tuist generate
```

then

```
xcodebuild … -derivedDataPath <custom path>
```

So xcodebuild will look for SwiftPM SourcePackages in this custom derived data path.

The problem is that the `tuist generate` step put the `SourcePackages` in the default DerivedData path ` /Users/bruno/Library/Developer/Xcode/DerivedData/<workspace-name>-<hash>`. Thus `xcodebuild` won't find the SPM packages that Tuist already resolved.

What I would like to do is:

```
DERIVED_DATA=<custom path> tuist generate
xcodebuild … -derivedDataPath <custom path>
```

so that both step work in sync.

**Comment by fortmarek** [comment]
Right, I see what you mean. I think we could create a new [generation option](https://docs.tuist.io/manifests/config#generationoption) such as `derivedDataPath: Path` that would count with the custom derived data path during the generation process as well. Would you like to implement this? As that is something not currently possible.

**Comment by bvirlet** [comment]
@fortmarek I'm willing to give it a try. If you have some insight about roughly what would need to be changed/where to add this option, that would be awesome.

I see the following actions:
- Option handling
- Passing option to xcodebuild command 

I suppose I can search the code base for `resolveDependenciesWithSystemScm` and the new option should be handled similarly.

**Comment by fortmarek** [comment]
Awesome!

Yes, you can basically follow that option and add `derivedDataPath` there as well, I don't think it differs much in how we should handle it.

**Comment by darrarski** [comment]
I need to customize `DerivedData` path for my project as well. While I don't have a solution for the problem, I did some research, and here is what I found:

Xcode allows customizing `DerivedData` location for a workspace. However, the configuration is stored in a user-specific location:

```
MyWorkspace.xcworkspace/xcuserdata/[username].xcuserdatad/WorkspaceSettings.xcsettings
```

Here is an example of the file with custom `DerivedData` location set:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>BuildLocationStyle</key>
	<string>UseAppPreferences</string>
	<key>CustomBuildLocationType</key>
	<string>RelativeToDerivedData</string>
	<key>DerivedDataCustomLocation</key>
	<string>DerivedData</string>
	<key>DerivedDataLocationStyle</key>
	<string>WorkspaceRelativePath</string>
	<key>IssueFilterStyle</key>
	<string>ShowActiveSchemeOnly</string>
	<key>LiveSourceIssuesEnabled</key>
	<true/>
	<key>ShowSharedSchemesAutomaticallyEnabled</key>
	<true/>
</dict>
</plist>
```

There seems to be another `WorkspaceSettings.xcsettings` file, stored in a shared location:

```
MyWorkspace.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings
```

Unfortunately, the custom `DerivedData` location stored in the shared file is ignored by Xcode :-(

I found that the `WorkspaceSettings.xcsettings` file can be generated by `XcodeProj` tool. [Here is a link](https://github.com/tuist/XcodeProj/blob/5964bb97b793bc0bd9d7fdf4cf66e59bcda4e19b/Sources/XcodeProj/Project/WorkspaceSettings.swift) to the source code file that I think is responsible for this. It won't allow customizing `DerivedData` location at the moment, but adding support for it should be straightforward. Unfortunately, I think it's pointless, as Xcode only respects this setting when it's stored in a user-specific `WorkspaceSettings.xcsettings` file. 

With this being said, I am not sure how to proceed, or if this is even possible to set a custom `DerivedData` path for the workspace and store this configuration in a way so it's shared.
